src/EventStore.Core/Messaging/TcsEnvelope.cs

```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using System.Threading.Tasks;
+
+namespace EventStore.Core.Messaging {
+	public class TcsEnvelope<TResult> : IEnvelope where TResult : class {
+		private readonly TaskCompletionSource<TResult> _tcs;
+		public Task<TResult> Task => _tcs.Task;
+
+		public TcsEnvelope() {
+			_tcs = new TaskCompletionSource<TResult>(TaskCreationOptions.RunContinuationsAsynchronously);
+		}
+
+		public void ReplyWith<T>(T message) where T : Message {
+			if (message is not TResult result)
```

Comment:
we it intentional to move the exception to being thrown on the one sending the reply? we could consider TrySetException

Reply:
fixed :+1: